File: codeK(main)/100 Day Challenge/Day 17/longestSubstring.java
//Longest substring without repeating characters - Leetcode question
import java.util.HashMap;

class LongestSubstring {
    public static void main(String[] args) {
        String s = "abcabcbb";
        int result = lengthOfLongestSubstring(s);
        System.out.println(result);
    }
    public static int lengthOfLongestSubstring(String s) {
        int maxLength = 0;
        HashMap<Character, Integer> stringMap = new HashMap<>();//new HashMap
        //s = {a,b,c,a,b,c,b,b}
        for (int i = 0; i < s.length(); i++) { //loops through the string
            if (stringMap.containsKey(s.charAt(i))) { //Checks if the char is already in the HashMap
                return maxLength = i - stringMap.get(s.charAt(i)); //returns current value of i - (index of the already initial char)
            } else {
                stringMap.put(s.charAt(i), i);//if not repeated, adds the char to the map
            }
        }
        return maxLength; // Returns the max length of the substring
    }
}

File: codeKenya S2/Day 5/Lynn Kamau/LongestSubstring.py
# Assignment 5: 3. Longest Substring Without Repeating Characters
# Given a string s, find the length of the longest substring without repeating characters.

def lengthOfLongestSubstring(s: str) -> int:
    current_substring = ""
    max_length = 0

    for char in s:
        if char in current_substring:
            index = current_substring.index(char)
            current_substring = current_substring[index + 1:]
        current_substring += char
        if len(current_substring) > max_length:
            max_length = len(current_substring)

    return max_length

# Example usage:
if __name__ == "__main__":      
    print(lengthOfLongestSubstring("abcabcbb"))  
    print(lengthOfLongestSubstring("pwwkew"))      
    print(lengthOfLongestSubstring("bbbbbb"))      
    print(lengthOfLongestSubstring(""))            